crates/red_knot_python_semantic/src/types.rs

```diff
@@ -401,27 +401,35 @@ impl<'db> Type<'db> {
     ///     pass
     /// ```
     ///
-    /// Returns `None` if `self` represents a type that is not iterable.
-    fn iterate(&self, db: &'db dyn Db) -> Option<Type<'db>> {
-        // `self` represents the type of the iterable;
-        // `__iter__` and `__next__` are both looked up on the class of the iterable:
-        let type_of_class = self.to_meta_type(db);
-
-        let dunder_iter_method = type_of_class.member(db, "__iter__");
-        if !dunder_iter_method.is_unbound() {
-            let iterator_ty = dunder_iter_method.call(db)?;
-            let dunder_next_method = iterator_ty.to_meta_type(db).member(db, "__next__");
-            return dunder_next_method.call(db);
-        }
+    /// Emits a diagnostic and returns `Unknown` if `self` represents a type that is not iterable.
+    fn iterate(&self, db: &'db dyn Db, mut diagnostic_fn: impl FnMut()) -> Type<'db> {
```

Comment:
I prefer to avoid passing callback functions whenever possible. It tends to lead to more verbose code and monomorphization. 

I wonder if we could use an approach similar to BiomeJS [`ParsedSyntax`](https://github.com/MichaReiser/biome/blob/98deae6e2409c1f9904356f6a6106e25cf3a89d6/crates/biome_parser/src/parsed_syntax.rs#L31) which is similar to an `Option` but with domain-specific helper methods. 

For example, we could define

```rust
enum ResolvedType<'db> {
	Known(Type<'db>),
	Unknown
}

impl<'db> ResolvedType<'db> {
	pub fn unwrap_with_diagnostic(self, builder: &mut TypeInferenceBuilder, node: AnyNodeRef, rule: &str, message: std::fmt::Arguments) -> Type<'db> {
		match self {
			Known(ty) => ty,
			Unknown(ty) => {
				builder.add_diagnostic(node, rule, message);
				Type::Unknown
			}
 	}
  }
}
```

Reply:
An alternative is that `unwrap_with_diagnostic` accepts an error builder similar to biomejs. This is fine because the `unwrap_with_diagnostic` function contains very little code. 

This is how it was used in BiomeJS:

```rust
parse_any_expression(p).or_add_diagnostic(p, expected_expression);
```

Reply:
Nice, I like the idea of the `ResolvedType` enum a lot!